### PR TITLE
Eve-7: Add example of calorimeter representation  of REveDataCollection

### PR DESCRIPTION
**Pull request by alja**
Add an example of a generic collection that is represented calorimeter towers. Implement selection so that REveDataCollection items in the table are consistent with CaloTower highlights.

Screen Shot 2020-11-09 at 2 12 10 PM

### CODE REVIEW

**Comment by alja** [comment]
@linev Please check.

**Comment by linev** [comment]
@phsft-bot build

File: graf3d/eve7/src/REveCaloData.cxx [hunk 1]
```diff
@@ -134,22 +134,25 @@ REveCaloData::REveCaloData(const char* n, const char* t):
...
-      secondary_idcs.insert(id);
+      fSelector->ProcessSelection(sel_cells, selectionId, multiple);
    }
+   else
+   {
+      REveSelection* selection = (REveSelection*) ROOT::Experimental::gEve->FindElementById(selectionId);
 
-   selection->NewElementPicked(GetElementId(), multiple, true, secondary_idcs);
+      std::set<int> secondary_idcs;
+      for (vCellId_i i = sel_cells.begin(); i != sel_cells.end(); ++i)
```
Comment (linev) [inline graf3d/eve7/src/REveCaloData.cxx:149]: Also here: `for (auto &i : sel_cells)`

File: tutorials/eve7/collection_proxies.C [hunk 2]
```diff
@@ -83,30 +96,121 @@ public:
...
+   std::vector<SliceInfo> fSliceInfos;
+   int fActiveSlice {-1};
+
+public:
+   REveCaloTowerSelector(REveCaloDataHist* cd) : fCaloData(cd){}
+   void AddSliceInfo(REveDataCollection* c, int s) { fSliceInfos.push_back(SliceInfo(c,s)); }
+   void SetActiveSlice(int a) { fActiveSlice = a; }
+
+   using REveCaloDataSelector::ProcessSelection;
+   virtual void ProcessSelection(REveCaloData::vCellId_t& sel_cells, UInt_t selectionId, Bool_t multi)
```
Comment (linev) [inline tutorials/eve7/collection_proxies.C:132]: Use `override` instead of `virtual`?

File: tutorials/eve7/collection_proxies.C [hunk 3]
```diff
@@ -83,30 +96,121 @@ public:
...
+                tower->fPhi > cd.fPhiMin && tower->fPhi < cd.fPhiMax)
+               item_set.insert(t);
+         }
+      }
+      REveSelection* sel = (REveSelection*)eveMng->FindElementById(selectionId);
+      sel->NewElementPicked(collection->GetItemList()->GetElementId(),  multi, true, item_set);
+   }
+
+   using REveCaloDataSelector::GetCellsFromSecondaryIndices;
+   virtual void GetCellsFromSecondaryIndices(const std::set<int>& idcs, REveCaloData::vCellId_t& out)
```
Comment (linev) [inline tutorials/eve7/collection_proxies.C:161]: Also here `override`

File: tutorials/eve7/collection_proxies.C [hunk 4]
```diff
@@ -83,30 +96,121 @@ public:
...
+      SliceInfo(REveDataCollection* c, int s) :fCollection(c), fSliceIdx(s) {}
+   };
+private:
+   REveCaloDataHist* fCaloData;
+   std::vector<SliceInfo> fSliceInfos;
+   int fActiveSlice {-1};
+
+public:
+   REveCaloTowerSelector(REveCaloDataHist* cd) : fCaloData(cd){}
+   void AddSliceInfo(REveDataCollection* c, int s) { fSliceInfos.push_back(SliceInfo(c,s)); }
```
Comment (linev) [inline tutorials/eve7/collection_proxies.C:128]: Use `fSliceInfos.emplace_back(c,s);`

**Review by linev — CHANGES_REQUESTED** [review]

File: graf3d/eve7/src/REveCalo.cxx [hunk 5]
```diff
@@ -941,7 +942,7 @@ void REveCalo2D::CellSelectionChangedInternal(REveCaloData::vCellId_t& inputCell
                if (!outputCellLists[bin])
                   outputCellLists[bin] = new REveCaloData::vCellId_t();
 
-               outputCellLists[bin]->push_back(REveCaloData::CellId_t((*i).fTower, (*i).fSlice, (*i).fFraction));
+               outputCellLists[bin]->push_back(REveCaloData::CellId_t((*i).fTower, (*i).fSlice, (*j).fFraction));
```
[the remark this answers is not in the document]
Comment (alja) [reply graf3d/eve7/src/REveCalo.cxx]: Thanks for info. It is changed in commit 7d71449.

File: graf3d/eve7/src/REveCaloData.cxx [hunk 6]
```diff
@@ -134,22 +134,25 @@ REveCaloData::REveCaloData(const char* n, const char* t):
...
+   if (fSelector)
    {
-      int id = (i->fSlice << 24) + i->fTower;
-      // printf("set set unique ID (%d, %d) -> [%d]\n",i->fSlice,  i->fTower, id);
-      secondary_idcs.insert(id);
+      fSelector->ProcessSelection(sel_cells, selectionId, multiple);
    }
+   else
+   {
+      REveSelection* selection = (REveSelection*) ROOT::Experimental::gEve->FindElementById(selectionId);
```
[the remark this answers is not in the document]
Comment (alja) [reply graf3d/eve7/src/REveCaloData.cxx]: Done!

File: graf3d/eve7/src/REveCaloData.cxx (hunk 1 above)
Comment (alja) [reply graf3d/eve7/src/REveCaloData.cxx:149]: @linev I have left this code unchanged to make it more explicit.

File: tutorials/eve7/collection_proxies.C (hunk 3 above)
Comment (alja) [reply tutorials/eve7/collection_proxies.C]: @linev Done in 69d9279.

File: tutorials/eve7/collection_proxies.C (hunk 2 above)
Comment (alja) [reply tutorials/eve7/collection_proxies.C]: @linev Done in 69d9279.

File: tutorials/eve7/collection_proxies.C (hunk 4 above)
Comment (alja) [reply tutorials/eve7/collection_proxies.C]: @linev Done in c79845c

**Comment by linev** [comment]
@phsft-bot build

**Comment by linev** [comment]
@phsft-bot build

**Review by linev — APPROVED** [review]

**Pull request merged**